Keep HTTP auth credentials when site is fireproof(ed)

aitorvs opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/414730916066338/1199538898514852/f
Tech Design URL: https://app.asana.com/0/0/1199538898514868/f
CC: 

**Description**:
This task fixes the issue that makes the HTTP auth credentials being removed upon firing even when the site is fireproofed.

As the `httpauth` table db where the credentials are stored is never read by Android, we can manually read/write/delete at will without interfering with Android.
We wrap the DB with Room just as a convenience.


**Steps to test this PR**:
Verify fireproofing works with migration
* install the app from develop branch
* visit https://httpbin.org/basic-auth/test/1111
* enter user = test, password = 1111
* fire
* visit https://httpbin.org/basic-auth/test/1111 
* verify the Auth dialog is shown again (credentials have been deleted)
* enter user = test, password = 1111
* upgrade the app from this branch
* launcher app and visit https://httpbin.org/basic-auth/test/1111 
* fire
* visit https://httpbin.org/basic-auth/test/1111 
* verify the Auth dialog is shown again (credentials have been deleted)
* enter user = test, password = 1111
* fireproof the site
* fire
* visit https://httpbin.org/basic-auth/test/1111 
* verify auth dialog is no longer shown and user is already authenticated

Verify fireproofing works with clean build
* Clean install the app from this branch
* visit https://httpbin.org/basic-auth/test/1111 
* enter user = test, password = 1111
* fire
* visit https://httpbin.org/basic-auth/test/1111 
* verify the Auth dialog is shown again (credentials have been deleted)
* enter user = test, password = 1111
* fireproof the site
* fire
* visit https://httpbin.org/basic-auth/test/1111 
* verify auth dialog is no longer shown and user is already authenticated

Verify credentials can be updated
* install the app from this branch
* visit https://httpbin.org/basic-auth/test/1111 
* enter user = test, password = 1111
* close and re-open the app
* visit https://httpbin.org/basic-auth/test/1111 
* verify user is logged in
* visit https://httpbin.org/basic-auth/test/2222
* verify auth dialog appears
* enter user = test, password = 2222
* verify user is logged in
* visit https://httpbin.org/basic-auth/test/1111 
* verify auth dialog appears

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/httpauth/db/HttpAuthEntity.kt:
```diff
@@ -0,0 +1,30 @@
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+// CREATE TABLE httpauth (_id INTEGER PRIMARY KEY, host TEXT, realm TEXT, username TEXT, password TEXT, UNIQUE (host, realm) ON CONFLICT REPLACE);
```
Question: how probable is it that WebView changes the schema of this database? is this schema defined in their codebase? if so, could we check if they have changed any time in the past?

aitorvs replied:
The documentation says webview never reads from this db.
The [source code](https://github.com/chromium/chromium/blame/master/android_webview/java/src/org/chromium/android_webview/HttpAuthDatabase.java) for the creation of the DB has been untouched for 8 years.

I think it's unlikely it'll change however we could always just migrate the entire DB and use our own to ensure the system does not touch it.

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt:
```diff
@@ -215,15 +217,17 @@ class BrowserWebViewClient(
 
                     if (credentials != null) {
-                        handler.proceed(credentials[0], credentials[1])
+                        handler.proceed(credentials.username, credentials.password)
```
👍 I like this change.

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/httpauth/db/WebViewStoreExtensions.kt:
```diff
@@ -0,0 +1,30 @@
+    clearFormData()
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+        val webViewDatabase = WebViewDatabase.getInstance(this.context)
+        webViewDatabase.clearFormData()
```
NIT: could we suppress this warning since we already have the API version guard?

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/WebDataManager.kt:
```diff
@@ -83,16 +83,10 @@ class WebViewDataManager @Inject constructor(
-    private fun clearAuthentication(webViewDatabase: WebViewDatabase) {
-        webViewDatabase.clearHttpAuthUsernamePassword()
+    private fun clearAuthentication(webView: WebView) {
+        GlobalScope.launch(Dispatchers.IO) {
```
why is this triggered as a new global coroutine? That will return without ensuring it was completed.

aitorvs replied:
Yes, I will follow the same approach I did for other methods and remove the `GlobalScope` here

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/httpauth/WebViewHttpAuthStore.kt:
```diff
@@ -0,0 +1,128 @@
+}
+
+class RealWebViewHttpAuthStore(
+    private val dispatcherProvider: DispatcherProvider,
```
This is currently not used, I guess we want to use it instead of `Dispatchers.IO`?

aitorvs replied:
Yes, good catch. That was the intention, will do

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/httpauth/WebViewHttpAuthStore.kt:
```diff
@@ -0,0 +1,128 @@
+class RealWebViewHttpAuthStore(
+    private val dispatcherProvider: DispatcherProvider,
+    private val fireproofWebsiteDao: FireproofWebsiteDao,
+    private val httpAuthDao: HttpAuthDao? = null
```
is this default param required? I like that we have the logic inside the `DatabaseModule` to decide why it's `null`, and I would expect the same from any place that creates an instance of this class.

aitorvs replied:
I can remove the `=null` yep

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/httpauth/WebViewHttpAuthStore.kt:
```diff
@@ -0,0 +1,128 @@
+
+@VisibleForTesting
+fun WebView.setHttpAuthUsernamePasswordCompat(host: String, realm: String, username: String, password: String) {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```
This is the same logic we have in `DatabaseModule`. I wonder if having it duplicated here without a clear connection between them can lead to possible bugs if something changes. is there any way to unify them or to connect them somehow? 

Checking this twice, it's like we have 2 approaches for the same. 
a) compat method that handles Android api version
b) we call this method after checking that `httpAuthDao` was null, and `httpAuthDao` is `null` based on the device Android Api version.

aitorvs replied:
This was on purpose. 
The webview extensions are disconnected from the DAO and viceversa. So we should not remove API checking from one of them assuming it will only be called from the other one and that other one will perform the checking.
If we remove the API checking from one of them and rely on the other one, we'd be coupling this two classes, something that I think is generally undesirable.

cmonfortep commented on app/src/androidTest/java/com/duckduckgo/app/browser/httpauth/WebViewHttpAuthStoreTest.kt:
```diff
@@ -0,0 +1,119 @@
+import org.junit.Test
+
+@ExperimentalCoroutinesApi
+class WebViewHttpAuthStoreTest {
```
There are a few warnings within this class that we may want to suppress.

aitorvs replied:
Yep, will do

cmonfortep commented on app/src/main/java/com/duckduckgo/app/di/DatabaseModule.kt:
```diff
@@ -29,6 +36,23 @@ import javax.inject.Singleton
+                .build()
+        } else null
+
+        return RealWebViewHttpAuthStore(DefaultDispatcherProvider(), fireproofWebsiteDao, httpAuthDb?.httpAuthDao())
```
Should we create this instance using an injected Dispatcher provider?

aitorvs replied:
It's s good question, I saw that all usages of `DefaultDispatcherProvider` are not through DI but rather building the type directly. That's why I didn't want to provide it, to avoid confusion.
As it is right now, it is still an injected dependency, just it is not available in the DI graph.
I think it is worth refactoring all places where we call `DefaultDispatcherProvider()` and get that thru DI but not sure that place is this PR.

cmonfortep replied:
Agree, I don't think we need to do that refactor within this PR. 

However, we already have `CoroutinesModule` that will provide that `DispatcherProvider` for you. And we have some Modules using it already (`UncaughtExceptionModule`, `BrowserModule`,...) So I would say we can use it in the same way.

aitorvs replied:
oh nice, I missed the fact that we had it already in the `CoroutinesModule`. Will do

cmonfortep commented on app/build.gradle:
```diff
@@ -148,6 +148,7 @@ dependencies {
     implementation "androidx.constraintlayout:constraintlayout:$constraintLayout"
     implementation "androidx.swiperefreshlayout:swiperefreshlayout:$swipeRefreshLayout"
     implementation "androidx.webkit:webkit:$webkit"
+    implementation "androidx.security:security-crypto:1.1.0-alpha03"
```
question: where are we using this dependency?

aitorvs replied:
we don't use it, my bad. It is a left over of something else I was doing. Removing it

cmonfortep reviewed (commented):
Great work! I really liked how you approached the problem and divided the different responsibilities. A few minor comments from my side. 

I've tested this on different devices and everything worked as expected.

aitorvs commented:
@cmonfortep thanks for the review. I have addressed all comments and and replied to [this one](https://github.com/duckduckgo/Android/pull/1066#discussion_r559648506)

aitorvs commented:
@cmonfortep `whenAddLoginDetectionThenJSInterfaceAdded` is failing for one of the bitrise devices. I was introduced in a previous PR. It is also failing locally, any clues? Maybe a flaky test?

cmonfortep reviewed (approved):
Thanks @aitorvs, great work! I see we managed to make this compatible for all api levels. Could we update the description? 

Apart from that, I've tested this on different devices and everything works great. :shipit:
